### ISSUE TITLE
Bug in parallel process

### ISSUE THREAD

**Issue by wilkerlucio** [wilkerlucio/pathom#198]
Reported on Clojurians

```clojure
(ns dev.dehli.pathom-bug
  (:require [cljs.core.async :refer [go <!]]
            [com.wsscode.pathom.core :as p]
            [com.wsscode.pathom.connect :as pc]))

(pc/defresolver other-key
  [{id :my/id}]
  {:my/other-key (str "hello, " id)})

(pc/defresolver state
  [{:keys [parser] :as env} {id :my/id}]
  {::pc/output [:my/id :my/state]}
  (go
    (let [ident [:my/id id]
          state (atom [])]
      ;; The 1st parse causes the 2nd parse call to return ::p/not-found
      ;; If you add :my/id to 1st query, the 2nd will work.
      (swap! state conj (<! (parser env [{ident [:my/other-key]}])))
      (swap! state conj (<! (parser env [{ident [:my/id]}])))
      {:my/state (mapv #(get % ident) @state)})))

(def parser
  (p/parallel-parser
    {::p/env     {::p/reader               [p/map-reader
                                            pc/parallel-reader
                                            pc/open-ident-reader
                                            p/env-placeholder-reader]
                  ::p/placeholder-prefixes #{">"}}
     ::p/plugins [(pc/connect-plugin {::pc/register [state
                                                     other-key]})
                  p/error-handler-plugin
                  p/trace-plugin]}))

(comment
  (go
    (prn (<! (parser {} [{[:my/id "foo"] [:my/state]}]))))
  ;;
  ;; {[:my/id "foo"] #:my{:state [#:my{:other-key "hello, foo"}
  ;;                              #:my{:id :com.wsscode.pathom.core/not-found}]}}
  )
```

**Comment by wilkerlucio** [comment]
The problem is that on the parallel parse, the entity gets merged and upon the second request the data for the ident is already there, and that causes the id to be missing, the following modified version highlights it:

```clj
(ns com.wsscode.pathom.debug
  (:require [clojure.core.async :refer [go <! <!!]]
            [com.wsscode.pathom.core :as p]
            [com.wsscode.pathom.connect :as pc]))

(pc/defresolver other-key
  [{id :my/id}]
  {:my/other-key (str "hello, " id)})

(pc/defresolver state
  [{:keys [parser] :as env} {id :my/id}]
  {::pc/output [:my/id :my/state]}
  (go
    (let [ident [:my/id id]]
      ;; The 1st parse causes the 2nd parse call to return ::p/not-found
      ;; If you add :my/id to 1st query, the 2nd will work.
      {:my/state [(p/entity env)
                  (<! (parser env [{ident [:my/other-key]}]))
                  (p/entity env)
                  (<! (parser env [{ident [:my/id]}]))]})))

(def parser
  (p/parallel-parser
    {::p/env     {::p/reader               [p/map-reader
                                            pc/parallel-reader
                                            pc/open-ident-reader
                                            p/env-placeholder-reader]
                  ::p/placeholder-prefixes #{">"}}
     ::p/plugins [(pc/connect-plugin {::pc/register [state
                                                     other-key]})
                  p/error-handler-plugin
                  p/trace-plugin]}))

(comment
  (<!! (parser {} [{[:my/id "foo"] [:my/state]}]))
  ;; {[:my/id "foo"] #:my{:state [#:my{:id "foo"}
  ;                             {[:my/id "foo"] #:my{:other-key "hello, foo"}}
  ;                    ;;;;;;;;; note here, [:my/id "foo"] is not part of the source data, so pathom wont add the id
  ;                             {:my/id "foo",
  ;                              [:my/id "foo"] #:my{:other-key "hello, foo"}}
  ;                             {[:my/id "foo"] #:my{:id :com.wsscode.pathom.core/not-found}}]}}
  )
```

I suggest you use a different approach to re-call the parser, instead of using idents, override the initial data, this will also make the output simpler to handle, as:

```clj
(ns com.wsscode.pathom.debug
  (:require [clojure.core.async :refer [go <! <!!]]
            [com.wsscode.pathom.core :as p]
            [com.wsscode.pathom.connect :as pc]))

(pc/defresolver other-key
  [{:keys [my/id]}]
  {:my/other-key (str "hello, " id)})

; small helper to reset entity
(defn with-entity [env ent]
  (assoc env ::p/entity (atom ent)))

(pc/defresolver state
  [{:keys [parser] :as env} {:keys [my/id]}]
  {::pc/output [:my/state]}
  (go
    ;;                           ;; override entity
    {:my/state [(<! (parser (with-entity env {:my/id id}) [:my/other-key]))
                (<! (parser (with-entity env {:my/id id}) [:my/id]))]}))

(def parser
  (p/parallel-parser
    {::p/env     {::p/reader               [p/map-reader
                                            pc/parallel-reader
                                            pc/open-ident-reader
                                            p/env-placeholder-reader]
                  ::p/placeholder-prefixes #{">"}}
     ::p/plugins [(pc/connect-plugin {::pc/register [state
                                                     other-key]})
                  p/error-handler-plugin
                  p/trace-plugin]}))

(comment
  (<!! (parser {} [{[:my/id "foo"] [:my/state]}]))
  )
```

**Comment by dehli** [comment]
Hello, I wanted to follow up with some additional details related to this bug. I've updated my code to utilize `with-entity`, however I'm still seeing an issue where occasionally parsers are executed without the full context. This bug is shown in the following code:

```clj
(ns dev.dehli.pathom-bug
  (:require [clojure.core.async :refer [go <! <!!]]
            [com.wsscode.pathom.core :as p]
            [com.wsscode.pathom.connect :as pc]
            [com.wsscode.pathom.parser :as pp]))

(defn with-entity [env ent]
  (prn (p/entity env))
  (assoc env ::p/entity (atom ent)))

(pc/defresolver one
  [{id :id}]
  {:one (str "one: " id)})

(pc/defresolver state
  [{:keys [parser] :as env} {id :id}]
  {::pc/output [:state]}
  (go
    (let [state (atom nil)]
      (reset! state (<! (parser (with-entity env {:id id}) [:one])))
      {:state @state})))

(def parser
  (p/parallel-parser
    {::p/env     {::p/reader               [p/map-reader
                                            pc/parallel-reader
                                            pc/open-ident-reader
                                            p/env-placeholder-reader]
                  ::p/placeholder-prefixes #{">"}
                  ::pp/external-wait-ignore-timeout 100}
     ::p/plugins [(pc/connect-plugin {::pc/register [state
                                                     one]})
                  p/error-handler-plugin
                  p/trace-plugin]}))

(comment
  (<!! (parser {} [{[:id "foo"] [:one :state]}]))

  ;; Returns
  ;; {[:id "foo"] {:one "one: foo" :state}}}
  )
```

By increasing `::pp/external-wait-ignore-timeout`, the query will take longer before returning empty state.
